charmcraft/commands/store/registry.py

```diff
@@ -440,3 +448,77 @@ def _upload_blob(self, filepath, size, digest):
 
         # finally remove the temp filepath
         os.unlink(filepath)
+
+    def upload_from_local(self, digest: str) -> str:
+        """Upload the image from the local registry."""
+        dockerd = LocalDockerdInterface()
+
+        # validate the image is present locally
+        logger.debug("Checking image is present locally")
+        image_info = dockerd.get_image_info(digest)
+        if image_info is None:
+            return
+        local_image_size = image_info["Size"]
+
+        logger.debug("Getting the image from the local repo; size=%s", local_image_size)
+        response = dockerd.get_streamed_image_content(digest)
+
+        tmp_exported = tempfile.NamedTemporaryFile(mode="wb")
+        extracted_total = 0
+        for chunk in response.iter_content(2 ** 20):
+            extracted_total += len(chunk)
+            progress = 100 * extracted_total / local_image_size
+            print("Extracting... {:.2f}%\r".format(progress), end="", flush=True)
+            tmp_exported.file.write(chunk)
+
+        # open the image tar and inspect it to get the config and layers for the manifest
+        tmp_exported.file.flush()
+        image_tar = tarfile.open(tmp_exported.name)
+        tmp_exported.close()  # closing implies deletion, but the tar lib already grabbed it ok
+        local_manifest = json.load(image_tar.extractfile("manifest.json"))
+        (local_manifest,) = local_manifest
```

Comment:
I'm curious about this line, why do we need this transformation? What is the original format of `local_manifest`?

Reply:
It's a list of one. Real life example:

```
[{'Config': '4616ab87e79ca4a4475211cf3f69eb40267c3b041f45b5bcecade8191fe2d97e.json',
  'Layers': ['fe44bcd0154eabf7a5ffdf2e270de700ea90c6155a7ccbd26bd8a791ce7e8e42/layer.tar',
             'bceaa585b938da41b8dfa56f497ece3a251cc7a70d9dde885499795b492f21eb/layer.tar',
             '7c46cbc5d4f29498503431edb7a151b4059a5e2ceaf1faf27a5d936198eb0213/layer.tar',
             'f615336d566d20f7bd51f01e28ad41bc7dc745e468eede82524a0508a8caf20f/layer.tar',
             '1e60b1df8ed245cbb286dd9e04f769a583534c0c396c583080e52c41dbad49b5/layer.tar',
             '0588cf2dc6457432fc6c2623f32f48fa82c4f6fd528d6296ad4dc74b8196a6bf/layer.tar',
             'e949573175f54ca50a024487a2126202e1e0217a8a88c8d027da17c9d10660b9/layer.tar',
             'db065a320c5e41441dfd3d58ea54ac1ba97a353becd925a7e30988854c18974a/layer.tar',
             '5728c11f035ae937880b6775817122890e0496bd60ad42de7967cbc08c58b2e8/layer.tar',
             '968d68636c32452e8ab9b2a401742fd65000eb91e689b879e4b128ccc20cc5be/layer.tar',
             'bfd82e46ae0b90aaadae7ba2a7dd72899fd2e63207f3165ff0cd45dfb49de0a7/layer.tar',
             '5e23da1d3ad0f4de734c74dd1fc615fc6ed88407638c9cc866582bfd289610c3/layer.tar'],
  'RepoTags': None}]
```

Improved the comment to reflect this.